Configurable inactivity timeout in console

jhadvig opened the pull request:
@spadgett @jwforres here is the first draft on the automatic logout upon user inactivity.
I've hardcoded the the inactivity period to 10 seconds, after that a dialog will be shown. If user cancels or dismisses the dialog then the inactivity period will be refreshed. If user wont make any activity or clicks 'Log Out' all the openshift tabs will be logged out. 

DEMO:
![nov 20 2017 4_33 pm](https://user-images.githubusercontent.com/1668218/33026999-347d8324-ce12-11e7-809b-c181319247fc.gif)

Will do additional refactoring next couple of days.

jwforres commented on app/views/modals/logout.html:
```diff
@@ -0,0 +1,13 @@
+      <h2>You are going to be logged out in <strong>{{timeToLogout}}</strong></h2>
+    </div>
+    <div class="modal-footer">
+      <button class="btn btn-lg btn-danger" type="submit" ng-click="logout();">Log Out</button>
```
I'm not sure what the appropriate order if for these buttons.  @ncameronbritt can you look into what the typical behavior is for these automated logout dialogs.  I don't want someone to panic and click logout when they meant to cancel the logout.

ncameronbritt replied:
Not sure you need two buttons. The action should be "stay logged in".  The user will be logged out anyway if she does nothing, hence this dialog. So is there any compelling reason to offer a "log out" action? I think it could be as simple as something like this: 
<img width="563" alt="screen shot 2017-11-22 at 8 54 47 am" src="https://user-images.githubusercontent.com/8974593/33130885-c68b941c-cf62-11e7-9a96-092606850ec9.png">

Another alternative is that you don't need any buttons. I like what they do [here](https://www.cloudera.com/documentation/enterprise/5-7-x/topics/cm_auto_logout.html) with a kind of "press any key to continue" prompt. If you're being logged out due to inactivity, it makes sense that any activity should keep the user logged in. 

Another thing I like about the example in that link is the way the log in page tells the user why they're seeing it again: "You were logged out due to inactivity."

cc @openshift/team-ux-review

jwforres commented on app/scripts/app.js:
```diff
@@ -481,6 +481,68 @@ angular
+    updateLogOut(false);
+    updateCloseAllModals(false);
+
+    $(document).bind("click keydown keypress", _.throttle(updateLastInteraction, 500));
```
why both keydown and keypress?

jwforres commented on app/scripts/app.js:
```diff
@@ -481,6 +481,68 @@ angular
+        showLogoutModal();
+      }
+      if (localStorage.getItem('logOut') === 'true') {
+        logout();
```
I'm trying to decide if its worth tracking this to force all the other tabs to go to the logout page immediately... the token is no longer valid as soon as any of the tabs go through logout. So any API interaction happening in the background on the other tabs would trigger a redirect to the login flow, and if no API interaction happened then they would just get to the logout page on their own when their own timer fired. 

@spadgett thoughts?

jhadvig replied:
Maybe just change the the message in the dialog, after the timeout.. something like:

![login](https://user-images.githubusercontent.com/1668218/33209573-c488d2be-d116-11e7-8dae-b85f99672c59.png)

That would make the logic more simple and we will solve the "redirect and notify" user issue, cause user will see that he was logged out.
@jwforres @ncameronbritt @serenamarie125  thoughts ?

jwforres commented on app/scripts/app.js:
```diff
@@ -481,6 +481,68 @@ angular
+      localStorage.setItem('lastInteraction', new Date());
+    };
+    var updateLogOut = function(bool) {
+      localStorage.setItem('logOut', bool);
```
logOut and closeAllModals aren't specific enough keys, even if you append origin-web-console- to them, it needs to be something like origin-web-console-inactivity-logout or something

that said, im not sure i want to track either of these, see other comments below

jwforres commented:
We should still have a button for mobile I would think

On Nov 22, 2017 9:05 AM, "Cameron Britt" <notifications@github.com> wrote:

> *@ncameronbritt* commented on this pull request.
> ------------------------------
>
> In app/views/modals/logout.html
> <https://github.com/openshift/origin-web-console/pull/2528#discussion_r152572964>
> :
>
> > @@ -0,0 +1,13 @@
> +<div class="modal-resource-action">
> +  <!-- Use a form so that the enter key submits when typing a project name to confirm. -->
> +  <form>
> +    <div class="modal-body">
> +      <h1>You have been inactive for period of time.</h1>
> +      <h2>You are going to be logged out in <strong>{{timeToLogout}}</strong></h2>
> +    </div>
> +    <div class="modal-footer">
> +      <button class="btn btn-lg btn-danger" type="submit" ng-click="logout();">Log Out</button>
>
> Not sure you need two buttons. The action should be "stay logged in". The
> user will be logged out anyway if she does nothing, hence this dialog. So
> is there any compelling reason to offer a "log out" action? I think it
> could be as simple as something like this:
> [image: screen shot 2017-11-22 at 8 54 47 am]
> <https://user-images.githubusercontent.com/8974593/33130885-c68b941c-cf62-11e7-9a96-092606850ec9.png>
>
> Another alternative is that you don't need any buttons. I like what they
> do here
> <https://www.cloudera.com/documentation/enterprise/5-7-x/topics/cm_auto_logout.html>
> with a kind of "press any key to continue" prompt. If you're being logged
> out due to inactivity, it makes sense that any activity should keep the
> user logged in.
>
> Another thing I like about the example in that link is the way the log in
> page tells the user why they're seeing it again: "You were logged out due
> to inactivity."
>
> cc @openshift/team-ux-review
> <https://github.com/orgs/openshift/teams/team-ux-review>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2528#discussion_r152572964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7foQHtMjA5JHFtLTHnwOU4YO3z_Pks5s5CpDgaJpZM4Qke-S>
> .
>

serenamarie125 commented:
Just had a conversation with a bunch of UX folks ... we landed here with the design: 
<img width="464" alt="screen shot 2017-11-22 at 10 25 14 am" src="https://user-images.githubusercontent.com/12504403/33135258-e75aca8e-cf6f-11e7-8b43-f9fcefd3ff6b.png">

The Continue Session button should be the default button, and the Log Out should not be red.

serenamarie125 commented:
@jwforres after being logged out, can we present the Login Screen with an inline notification (warning) stating "You have been logged out due to inactivity"

ncameronbritt commented:
@jhadvig @jwforres Related to this, I opened this issue: https://github.com/openshift/origin-web-console/issues/2530

jwforres commented:
On our own logout page yeah we could. But there is a URL we optionally
redirect to afterward that can be configured by the cluster admin so that
they can logout of their SSO environment completely if that's what they
want. That page is out of our hands at that point.

On Nov 22, 2017 10:36 AM, "Serena Doyle" <notifications@github.com> wrote:

> @jwforres <https://github.com/jwforres> after being logged out, can we
> present the Login Screen with an inline notification (warning) stating "You
> have been logged out due to inactivity"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2528#issuecomment-346386578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7dXcPg1H_RvcLiAFPkNv310l7OdHks5s5D-KgaJpZM4Qke-S>
> .
>

jhadvig commented on app/scripts/app.js:
```diff
@@ -481,6 +481,68 @@ angular
+      if (localStorage.getItem('logOut') === 'true') {
+        logout();
+      }
+    }, 1000);
```
yeah, this was just a example value. doesnt make sense to check every second... If the timeout will be in minutes then it make sense to set the check for every minute.

jhadvig commented on app/scripts/controllers/modals/logoutModal.js:
```diff
@@ -0,0 +1,35 @@
+ */
+angular.module('openshiftConsole')
+  .controller('LogoutModalController', function ($timeout, $scope, $uibModalInstance) {
+  	$scope.timeToLogout = 20;
```
You mean like a constant in app.js ?

spadgett commented on app/scripts/app.js:
```diff
@@ -481,6 +481,68 @@ angular
+    }
+    var isLocalLogoutModalShown = false;
+    var updateLastInteraction = function() {
+      localStorage.setItem('lastInteraction', new Date());
```
We've been using the prefix `openshift/` for other keys and kebab case names, so I'd suggest something like `openshift/last-interaction-time`.

serenamarie125 commented:
@jhadvig @jwforres what is the inactivity time period being set to ?

spadgett commented:
@serenamarie125 It will be configurable (disabled by default).

serenamarie125 commented:
great thanks!

jhadvig commented:
@spadgett @jwforres I've update the PR addressing comments

The modal will now contain following btns:
![logout-modal](https://user-images.githubusercontent.com/1668218/34111482-2578b34a-e40a-11e7-9856-bc64e7a4ef2e.png)

When user is logout by inactivity he will see it in the logout page:
![logout-timeout](https://user-images.githubusercontent.com/1668218/34111528-445f3f36-e40a-11e7-80b2-d7a1cc0d615e.png)

The default logout page looks like:
![logout-def](https://user-images.githubusercontent.com/1668218/34111543-533811fe-e40a-11e7-872c-c3ece3a71d84.png)

Also I've made following variables configurable:
 - **is_active** - Set to true if the the console shall check for user inactivity.
 - **timeout_period**  -  Set to number of milliseconds after which a modal with countdown should appear.
 - **check_for_timeout_period** - Set to number of milliseconds that shall console check for timeout.
 - **modal_countdown_period** -  Set to number of milliseconds that will be used in a for visual countdown in the modal.

jhadvig commented:
@ncameronbritt @serenamarie125 PTAL as well

ncameronbritt commented:
I would change the first sentence in the modal to say "Your session is about to expire due to inactivity." (I.e. give the reason why the session is expiring.)

This is a little bit of a nit, but I think the message on the logout page reads a little better as "You have been logged out due to inactivity."

jhadvig commented:
@ncameronbritt comments addressed ! Thanks not it looks better :) :+1:

jhadvig commented on app/scripts/constants.js:
```diff
@@ -109,6 +109,17 @@ angular.extend(window.OPENSHIFT_CONSTANTS, {
     namespace: "openshift"
   },
 
+  INACTIVITY_TIMEOUT: {
```
@jwforres @spadgett here we need to align on the default time that will be set. Also I believe that the `is_active` variable should be set to **false**, right ?

spadgett replied:
There probably only needs to be one value. I'd suggest letting the user set the timeout in minutes. Make any value that is not positive (undefined, null, 0, -1) could mean off? @jwforres opinion?

Default should be disabled, yes.

Ultimately this should go into the AssetConfig when we split the console into a pod and openshift-ansible updated to have a flag to set it. But that can happen in a later PR.

jwforres replied:
Agreed, one variable. Any positive integer is on, anything else is off.

jhadvig replied:
So based on https://github.com/openshift/origin-web-console/pull/2528#discussion_r152555338 I thought that we want to be able to configure multiple varibles like those mentioned in https://github.com/openshift/origin-web-console/pull/2528#issuecomment-352446268, so we don't have to hardcode it into code.
In case we really want just a single variable that would set the inactivity period in minutes(none positive value would disable the inactivity logout), we need to align on following values that would need to be hardcoded:
**check_for_timeout_period** - Seconds or that shall console check for timeout - I guess it would be enough to check every minute ?
**modal_countdown_period** - Set to number of milliseconds that will be used in a for visual countdown in the modal - I've set it to 20 seconds in demo. Guess thats OK ?

jwforres replied:
sorry i think there is some confusion here. I was just saying that you don't need `is_active` since setting `timeout_period` to something other than a positive integer would indicate whether we should be checking a timeout or not. And I agree with @spadgett that the timeout period should be configured in minutes.

I don't think you need `check_for_timeout_period`, checking once a minute for inactivity is sufficient. We will just document that it works that way.

For `modal_countdown_period` I would default to 30 seconds, that is what I commonly see.  I would also configure it in seconds, not milliseconds.  @spadgett were suggesting that we didn't need this one?

spadgett replied:
> For modal_countdown_period I would default to 30 seconds, that is what I commonly see. I would also configure it in seconds, not milliseconds. @spadgett were suggesting that we didn't need this one?

Personally I don't see a ton of value in making it configurable and might cause some confusion in how the two values are different when an admin first configures it.

jwforres replied:
Ok i'm fine with that. Let's just make it 30 seconds then and we'll need to document it.

jhadvig commented:
@spadgett @jwforres comments addressed. Now we only have one configurable variable **INACTIVITY_TIMEOUT_PERIOD** which is defaulted to false, otherwise it takes amount of minutes.
PTAL

spadgett commented on app/scripts/controllers/modals/logoutModal.js:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    var logoutTimeout = $timeout(onTimeout,1000);
```
+1, I'd probably keep a timestamp for when the logout will happen and then use relative time directive in the dialog. I'm not sure you really need a $timeout every second here (but I haven't looked closely at the code).

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,80 @@ angular
+      localStorage.setItem('origin-web-console-inactivity-logout', bool);
+    };
+    var updateCloseAllModals = function(bool) {
+      localStorage.setItem('origin-web-console-dismiss-logout-modals', bool);
```
i think we can eliminate this local storage variable.  when the user clicks the button to dismiss the dialogs and remain active that is going to set the last interaction time.

you can then have your modal add a listener on local storage https://stackoverflow.com/questions/28230845/communication-between-tabs-or-windows

that listener can dismiss the modal if it sees the key for the last interaction time is updated

you probably want to make sure that clicking the logout link does not reset the last interaction time, stopping propagation of the event from the ng-click of the button would probably fix this.

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,80 @@ angular
+      localStorage.setItem('origin-web-console-last-interaction-timestamp', new Date());
+    };
+    var updateLogOut = function(bool) {
+      localStorage.setItem('origin-web-console-inactivity-logout', bool);
```
you need to make sure this localStorage gets reset on login, the auth service has something like an onUserChanged i think that you could listen to

jhadvig replied:
How about clearing the localStorage on logout ? Wouldnt that be enough ?

jwforres replied:
no because there may be other tabs that haven't yet seen the flag to logout, think if you have 3+ tabs open

jhadvig replied:
That would also handle the case you mentioned bellow 
> you probably want to make sure that clicking the logout link does not reset the last interaction time, stopping propagation of the event from the ng-click of the button would probably fix this.

jhadvig commented on app/scripts/app.js:
```diff
@@ -485,6 +485,81 @@ angular
+    }
+    var isLocalLogoutModalShown = false;
+    var updateLastInteraction = function() {
+      localStorage.setItem('origin-web-console-last-interaction-timestamp', new Date());
```
But the value of the latest interaction is set in different cases:
1, when the loading the page ,but it might happen that user wont interact with the page anymore so we need to set it by default
2, on user interaction
3, when dismissing the logout modal

Not really sure whats the benefit of setting a local var and updating the localStorage with it everytime, since the those cases might be mutual exclusive. Thought that having a simple helper that would just update the latestInteraction key in the local storage directly when any of those cases occur

jwforres replied:
I'm talking about the key string, not the value.

jwforres replied:
Treat it like a constant

jwforres replied:
As in `var lastInteractionKey = "origin-web-console-last-interaction-timestamp";`

spadgett replied:
You should be passing a string as the value to `setItem` not a `Date` object.

jhadvig commented:
@jwforres comments addressed. 
Added `countdownFromTimestamp` filter as suggested per our conversation. Also added comments to the functions so the code is more understandable. 
I've tested the changes and all the cases should be handled (single and also multi tab). Sorry that it took a bit longer. But there have been more edge cases then I've expected.
cc'ing @spadgett

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,110 @@ angular
+        }
+      }, function () {
+        updateLastInteraction();
+        restartCheckInteractionInterval();
```
because you call updateLastInteraction which updates the local storage, that means the event below is going to fire, so restartCheckInteractionInterval gets called twice.  Do you actually need it here?

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,110 @@ angular
+    // Reset 'origin-web-console-inactivity-logout' key on login.
+    AuthService.onUserChanged(logoutAllTabs(false));
+    updateLastInteraction();
+    startInteractionIntervalCheck();
```
same question as above, isnt the storage event going to fire and start the listener?

jwforres commented on app/scripts/controllers/modals/logoutModal.js:
```diff
@@ -0,0 +1,38 @@
+    var lastInteraction = Date.parse(localStorage.getItem('origin-web-console-last-interaction-timestamp'));
+    var countdownInterval = setInterval(function() {
+      var timeRemaining = $filter('countdownToTimestamp')($scope.endTimestamp);
+      $('.countdown').text(function() {
```
this shouldn't be in here, it should be in the same block in app.js that handles the other date / time counters

jhadvig replied:
Not really sure if what would be the benefit of having it in `app.js`. Isn't it better to have the have it separated inside the modal controller as part of the interval, since we have to update the DOM with the actual countdown number every second ?

jwforres commented on app/scripts/controllers/modals/logoutModal.js:
```diff
@@ -0,0 +1,38 @@
+      if (timeRemaining == 0) {
+        $scope.logout();
+      }
+    }, 1000);
```
instead of an interval here you just need a single timeout of 30 seconds that runs the logout, clear the timeout below instead of clear interval

jhadvig replied:
But the interval makes sure that the displayed countdown number will be count down, the same as in https://github.com/openshift/origin-web-console/blob/master/app/scripts/app.js#L491-L505

jhadvig commented:
@jwforres comments addressed. PTAL

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,118 @@ angular
+    };
+    // Updates lastInteraction date and restarts interval
+    var updateLastInteraction = function() {
+      restartCheckInteractionInterval();
```
looks like another case where restartCheckInteractionInteral doesnt need to be called directly since it will get called from the event

jhadvig replied:
We need to call it here because the listener on the `storage` event is not fired in the tab where the event happened.
Removing `updateLastInteraction()` from the the modal resolving block remove the cases where the `restartCheckInteractionInterval()` is called twice (tested it)

jwforres replied:
Ok good to know

jwforres commented on app/scripts/app.js:
```diff
@@ -485,6 +485,118 @@ angular
+    }
+
+    // Set interval for all timeRemainingFromNow directives.
+    setInterval(function() {
```
I don't want this in this block, since this block is not run unless the inactivity timeout feature is enabled.  The directive / filter are independent of that feature. Just take the content of this function and add it to the existing interval that is handling the duration timestamps

jwforres replied:
So I would expect it to be part of this same interval https://github.com/openshift/origin-web-console/blob/master/app/scripts/app.js#L504

jhadvig commented:
@jwforres comments addressed. It really makes sense to have countdown in a single interval :+1:

jwforres commented on app/scripts/constants.js:
```diff
@@ -109,6 +109,11 @@ angular.extend(window.OPENSHIFT_CONSTANTS, {
+
+  // Set to number of minutes after which a modal with countdown should appear.
+  // When set to anything else then positive integer the inactivity timeout won't be enable. 
+  INACTIVITY_TIMEOUT_PERIOD: 1,
```
Think this accidentally got checked in enabled

jhadvig replied:
ohh... was retesting and accidentally forgot to disable it :-/

jwforres commented:
Looks good now, can you add the dist files?

jhadvig commented:
@jwforres dist added

jwforres commented:
/lgtm
